app/forms/submissions/missing_data_validator.rb

```diff
@@ -16,7 +16,9 @@ def missing_data_validator(name, options)
     missing_data_validator :placements, form: "PlacementsForm", if: :requires_placements?
 
     def missing_fields
-      forms.map(&:missing_fields).flatten.uniq
+      forms.map(&:missing_fields).tap do |fields|
```

Comment:
It feels a bit strange to me having this logic here. Should the `missing_fields` method in the DegreesForm be the place to handle this?

Reply:
All the errors that would come from the `DegreesForm` need to be propagated to the `Submissions::MissingDataValidator` as the latter is been called at [trainee.rb#L604](https://github.com/DFE-Digital/register-trainee-teachers/blob/7308-prevent-qts-recommendation-being-made-in-register-ui-if-degree-information-has-not-been-added/app/models/trainee.rb#L604) which will determine the value of `trainee.submission_ready?` to be used at [award_recommendations_controller.rb#L6](https://github.com/DFE-Digital/register-trainee-teachers/blob/7308-prevent-qts-recommendation-being-made-in-register-ui-if-degree-information-has-not-been-added/app/controllers/trainees/award_recommendations_controller.rb#L6)

There is also the requirement to conditionally render the **Recommend trainee for QTS** button which depends on `Submissions::MissingDataValidator#missing_fields`.

Reply:
Doesn't the [missing_fields](https://github.com/DFE-Digital/register-trainee-teachers/blob/44fe582e0a6dc63ffeb2808818a66fe4c6ba5087/app/forms/degrees_form.rb#L82) method on DegreesForm do the same thing here anyway? 

```ruby
def missing_fields(include_degree_id: false)
    return [] if valid?

    degrees.map do |degree_form|
      degree_form.valid?

      if include_degree_id
        { degree_form.slug => degree_form.errors.attribute_names }
      else
        degree_form.errors.attribute_names
      end
    end
  end
```

That will return an array of attribute names for the errors on the form, like you have on line 20 here. So this method can just be: 
```ruby 
def missing_fields
  forms.map(&:missing_fields).flatten.uniq
end
``` 
and the behaviour is the same?

Reply:
Yeah, I guess it's unclear to me what the intention was for `missing_fields` on `DegreesForm` before these changes. Should it:
1. Only capture missing fields in Degrees that have been added, or
2. Do that, _and_ capture whether Degrees themselves are missing?

The way it's coded suggests number 1 but the way it's being called by `Submissions::MissingDataValidator` suggests it should be number 2.

Reply:
> Doesn't the [missing_fields](https://github.com/DFE-Digital/register-trainee-teachers/blob/44fe582e0a6dc63ffeb2808818a66fe4c6ba5087/app/forms/degrees_form.rb#L82) method on DegreesForm do the same thing here anyway?
> 
> ```ruby
> def missing_fields(include_degree_id: false)
>     return [] if valid?
> 
>     degrees.map do |degree_form|
>       degree_form.valid?
> 
>       if include_degree_id
>         { degree_form.slug => degree_form.errors.attribute_names }
>       else
>         degree_form.errors.attribute_names
>       end
>     end
>   end
> ```
> 
> That will return an array of attribute names for the errors on the form, like you have on line 20 here. So this method can just be:
> 
> ```ruby
> def missing_fields
>   forms.map(&:missing_fields).flatten.uniq
> end
> ```
> 
> and the behaviour is the same?

If we look at the `DegreesForm#missing_fields` if does the following:
1. return an empty Array in the case the form is valid
2. map the trainee degrees errors from `DegreesForm#degrees`

Number 2. won't run in the case that are no Degrees `[].map {}` therefore `DegreesForm#missing_fields` will return `[]`

> Yeah, I guess it's unclear to me what the intention was for `missing_fields` on `DegreesForm` before these changes. Should it:
> 
> 1. Only capture missing fields in Degrees that have been added, or
> 2. Do that, _and_ capture whether Degrees themselves are missing?
> 
> The way it's coded suggests number 1 but the way it's being called by `Submissions::MissingDataValidator` suggests it should be number 2.

I think the intention of `DegreesForm#missing_fields` is to validate the `DegreesForm#degrees`. Probably the intention is to allow the forms to be valid until a provider decides to add degrees to a trainee.

Reply:
> > Yeah, I guess it's unclear to me what the intention was for `missing_fields` on `DegreesForm` before these changes. Should it:
> > 
> > 1. Only capture missing fields in Degrees that have been added, or
> > 2. Do that, _and_ capture whether Degrees themselves are missing?
> > 
> > The way it's coded suggests number 1 but the way it's being called by `Submissions::MissingDataValidator` suggests it should be number 2.
> 
> I think the intention of `DegreesForm#missing_fields` is to validate the `DegreesForm#degrees`. It won't return any errors from the `DegreesForm` itself. Probably to allow the forms to be valid until a provider decides to add degrees to a trainee.

Yeah, that could be it, but it's difficult to know the intention. We may want to change this in the future anyway to better fit our needs. Consider the equivalent in PlacementsForm which gives different behaviour:

https://github.com/DFE-Digital/register-trainee-teachers/blob/main/app/forms/placements_form.rb#L110

Anyway, for now I think let's get this merged in. It works for our use case, and we can revisit in the future if we need to.